### ISSUE TITLE
UTF-8 characters getting problems

### ISSUE THREAD

**Issue by JiriSch** [barbushin/php-imap#657]
Thank you for this library, I am using it for years, I was using up to version 3.1.0 but now I wanted to upgrade to 4.5 ang get problems with UTF-8 characters.

 - PHP IMAP version: tested anything newver than 3.1.0
 - PHP Version: 7.4
 - Type of execution: web server

Email with UTF-8 characters - these characters are unreadable.

The used code:
```php
	$mailbox = new PhpImap\Mailbox('{xxxxxx:143/imap4/notls/novalidate-cert}INBOX', 'xxxxx', 'xxxx');
	var_dump($mailbox->getServerEncoding()); //---GETTING UTF-8---
	$mailsIds = $mailbox->searchMailbox('ALL');
	if(!$mailsIds) return 0;
	foreach ($mailsIds as $key=>$mailsId)
	{
		$mail = $mailbox->getMail($mailsId);
		//var_dump($mail);
		
		var_dump($mail->textHtml);
	}
```

I tried to test any version newer than 3.1.0 but without success, only till 3.1.0 works.

Thank you.

**Comment by JiriSch** [comment]
I add some more debug info, textPlain downloaded over 3.1.0:

`With interpunction: ěščřžýáíé
Without: escrzyaie`

The same email downloaded over any version newer, for ex 4.5.1:

`With interpunction: ��������
Without: escrzyaie`

In this attachment is the raw email from IMAP server (downloaded from IMAP server by SCP), sended by MS Outlook, recieved by Postfix, IMAP use DOVECOT:
[EmailTest.txt](https://github.com/barbushin/php-imap/files/7946228/EmailTest.txt)

**Comment by JiriSch** [comment]
I have fixed this issue by changing:

> foreach ($elements as $element) {
> $newString .= $this->convertToUtf8($element->text, $element->charset);
> }

to:
> foreach ($elements as $element) {
> if ($element->charset!='default') $this->decodeMimeStrDefaultCharset = $element->charset;
> $newString .= $this->convertToUtf8($element->text, $element->charset);
> }

in Mailbox.php:1490

Not sure if this is the right solution, but it works for me...

**Comment by JiriSch** [comment]
Just to explain, it looks like 

> $elements = \imap_mime_header_decode($string);

returns in $elements->charset default in case it's not, so this handle this from another elements...

**Comment by hrst** [comment]
> I tested it with quite a lot emails. Most iso-8859-1 emails are fixed with the above patch. However, I still found some iso-8859-1 emails, that get 'default' as charset from imap_mime_header_decode in decodeMimeStr.
> 
> Maybe it is better to fix it this way in DataPartInfo.php:111? At least this works for me with all tested emails.

I can confirm, this fixes encoding issues for me.

I think this library desperately needs unit tests or similar that load a set of test emails and then checks whether they encoded correctly. During the last versions, encoding issues that were [already fixed](https://github.com/barbushin/php-imap/issues/543) reappeared multiple times.

**Comment by Sebi94nbg** [comment]
Thanks for your code improvements, testings and feedback.

I've created a merge request with your changes, but I need to check, if we can implement it like this and if we instead can / need to remove something else as it doesn't make any sense anymore then.

Feel free to comment and provide feedback on the above linked merge request. :)

Yes, I agree, we need some unit tests, but I'm unfortunately not that familiar with writing unit tests, so I'm not really sure, how we can properly test this. Feel free to create a merge request with some or let me know, how we can define such a good unit test, then I'll add it to the above merge request.

**Comment by Sebi94nbg** [comment]
I've just released a new version [4.5.3](https://github.com/barbushin/php-imap/releases/tag/4.5.3) with the above fix and some unit tests.
